### PR TITLE
Remove json serialization annotations from internal types

### LINKED ISSUE #3933: Remove json tags from pkg/api/types.go
Previous discussion in #1004 and #3900.

We should never serialize the internal representation. Removing the json tags would make this clearer.

/cc @thockin

### PR DESCRIPTION
**Pull request by liggitt**
fixes #3933

Internal types should never be serialized, and including json serialization tags on them makes it possible to accidentally do that without realizing it.

fixes in this PR:

* types
  * [x] remove json tags from internal types
  * [x] fix references from serialized types to internal ObjectMeta
* generation
  * [x] remove generated json codecs for internal types (they should never be used)
* kubectl
  * [x] fix `apply` to operate on versioned object
  * [x] fix sorting by field to operate on versioned object
  * [x] fix `--record` to build annotation patch using versioned object
* hpa
  * [x] fix unmarshaling to internal CustomMetricTargetList in validation
* thirdpartyresources
  * [x] fix encoding API responses using internal ObjectMeta
* tests
  * [x] fix tests to use versioned objects when checking encoded content
  * [x] fix tests passing internal objects to generic printers

follow ups (will open tracking issues or additional PRs):
- [x] remove json tags from internal kubeconfig types (`kubectl config set` pathfinding needs to work against external type) - tracked in https://github.com/kubernetes/kubernetes/issues/5625
- [x] HPA should version CustomMetricTargetList serialization in annotations - https://github.com/kubernetes/kubernetes/pull/39407
- [x] add static analysis tests preventing new internal types from adding tags - https://github.com/kubernetes/kubernetes/pull/39152
- [x] add static analysis tests requiring json tags on external types (and enforcing lower-case first letter) - https://github.com/kubernetes/kubernetes/pull/39152

### CODE REVIEW

**Comment by liggitt** [comment]
cc @kubernetes/sig-api-machinery

**Comment by deads2k** [comment]
I want it, but I suspect it will be a painful trip. Did `apply` ever start doing better?

**Comment by smarterclayton** [comment]
100% supportive of this.

**Comment by caesarxuchao** [comment]
> Did apply ever start doing better?

`kubectl apply`? Doesn't it operate on the versioned objects?

**Comment by bgrant0607** [comment]
Thanks!!!!

Have the previously known blockers (e.g., jsonpath) already been eliminated?

**Comment by smarterclayton** [comment]
HA.  You'd think so, wouldn't you?

**Comment by bgrant0607** [comment]
cc @pwittrock @ymqytw re. apply

**Comment by bgrant0607** [comment]
cc @caesarxuchao @mbohlool re. client and OpenAPI issues

**Comment by liggitt** [comment]
apply and jsonpath are still using internal object json tags. will work through those as part of this (help appreciated)

**Comment by caesarxuchao** [comment]
Ref: `kubectl apply` operates on internal object: https://github.com/kubernetes/kubernetes/pull/35688.

File: pkg/kubectl/cmd/apply.go [hunk 1]
```diff
@@ -527,14 +527,11 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 		return nil, cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", obj), source, err)
 	}
 
-	// Create the versioned struct from the original from the server for
-	// strategic patch.
-	// TODO: Move all structs in apply to use raw data. Can be done once
-	// builder has a RawResult method which delivers raw data instead of
-	// internal objects.
-	versionedObject, _, err := p.decoder.Decode(current, nil, nil)
```
Comment (liggitt) [inline pkg/kubectl/cmd/apply.go]: @soltysh @smarterclayton thoughts on this? `p.decoder.Decode(current, nil, nil)` was returning internal objects before

File: pkg/registry/extensions/thirdpartyresourcedata/codec.go [hunk 2]
```diff
@@ -480,7 +480,11 @@ func encodeToJSON(obj *extensions.ThirdPartyResourceData, stream io.Writer) erro
 	if !ok {
 		return fmt.Errorf("unexpected type: %v", objOut)
 	}
-	objMap["metadata"] = obj.ObjectMeta
+
+	versionedObjectMeta := &v1.ObjectMeta{}
+	v1.Convert_api_ObjectMeta_To_v1_ObjectMeta(&obj.ObjectMeta, versionedObjectMeta, nil)
```
Comment (smarterclayton) [inline pkg/registry/extensions/thirdpartyresourcedata/codec.go]: .... twitch

File: pkg/kubectl/cmd/util/helpers.go [hunk 3]
```diff
@@ -523,18 +523,24 @@ func RecordChangeCause(obj runtime.Object, changeCause string) error {
 // ChangeResourcePatch creates a strategic merge patch between the origin input resource info
 // and the annotated with change-cause input resource info.
 func ChangeResourcePatch(info *resource.Info, changeCause string) ([]byte, error) {
-	oldData, err := json.Marshal(info.Object)
+	// Get a versioned object
+	obj, err := info.Mapping.ConvertToVersion(info.Object, info.Mapping.GroupVersionKind.GroupVersion())
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/helpers.go]: You may already have versionedObject set at this point - would probably prefer to use that unless someone earlier is mutating without clearing VersionedObject.

File: pkg/kubectl/resource/visitor.go [hunk 4]
```diff
@@ -81,9 +81,9 @@ type Info struct {
 	// Optional, this is the provided object in a versioned type before defaulting
 	// and conversions into its corresponding internal type. This is useful for
 	// reflecting on user intent which may be lost after defaulting and conversions.
-	VersionedObject interface{}
+	VersionedObject runtime.Object
```
Comment (smarterclayton) [inline pkg/kubectl/resource/visitor.go]: Hrm... I can't remember why this was interface{}

File: pkg/kubectl/cmd/util/helpers.go (hunk 3 above)
Comment (liggitt) [inline pkg/kubectl/cmd/util/helpers.go]: I thought versioned object came from the user, and object came from the server...

File: pkg/kubectl/cmd/apply.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/kubectl/cmd/apply.go]: Looks correct.

**Comment by liggitt** [comment]
green tests!

will rebase on https://github.com/kubernetes/kubernetes/pull/38451 once it merges and address comments, but this is ready for review

File: pkg/kubectl/cmd/util/helpers.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/helpers.go]: Bah, you're right.  I want to kill VersionedObject anyway so this is fine.

**Comment by liggitt** [comment]
@k8s-bot gci gke e2e test this

File: pkg/kubectl/sorting_printer.go [hunk 5]
```diff
@@ -103,7 +103,9 @@ func SortObjects(decoder runtime.Decoder, objs []runtime.Object, fieldInput stri
 		switch u := item.(type) {
 		case *runtime.Unknown:
 			var err error
-			if objs[ix], _, err = decoder.Decode(u.Raw, nil, nil); err != nil {
+			// decode runtime.Unknown to runtime.Unstructured for sorting.
+			// we don't actually want the internal versions of known types.
+			if objs[ix], _, err = decoder.Decode(u.Raw, nil, &runtime.Unstructured{}); err != nil {
```
Comment (smarterclayton) [inline pkg/kubectl/sorting_printer.go]: I don't suppose there is anything that tests this?  From david's generic get work?

Comment (liggitt) [inline pkg/kubectl/sorting_printer.go]: sorting_printer_test hits it pretty well... broke when I removed the json tags from internal

**Comment by liggitt** [comment]
@k8s-bot gci gke e2e test this

**Comment by smarterclayton** [comment]
LGTM

**Comment by smarterclayton** [comment]
Smoke failure is a flake

**Comment by smarterclayton** [comment]
Ahhh... so much cleaner now.

**Comment by liggitt** [comment]
tag tests added in https://github.com/kubernetes/kubernetes/pull/39152

**Pull request merged**